### PR TITLE
Prevent setting recipe to red belt/underground/splitter in asm from untrusted 

### PR DESCRIPTION
**Pull request by Ragnarok77-factorio**
### Brief description of the changes:
Prevent untrusted players from setting the recipe to asm to red belt/underground/splitter. It also prevents someone from copying a red belt/underground/splitter to another asm (so he can't work around that)

Method chosen  :
I used on gui closed as it was the most fps/ups friendly way to get this feature (another solution could be a on_tick that checks every once in a while a list of asms that is made every time an asm is built but that's not fps/ups friendly, but would prevent any exploit to dodge it). Sadly there is no on recipe change event

Drawback 1 :
One drawback is if the player alt f4, then he can dodge the on_gui_closed event and have his red asm (we could prevent that by checking on player leave if he made a red belt&co asm, and having the list of asm generated with on built event, but not that really fps/ups friendly), and not sure it's worth the overhead. As you want if you want to prevent that or not.

For a new player, I doubt he can find it to be honest so not sure it's an issue that needs to be resolved.

Vote : https://discord.com/channels/823696400797138974/823771211421974579/994646509700259970

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### CODE REVIEW

File: maps/biter_battles_v2/main.lua [hunk 1]
```diff
@@ -31,6 +31,19 @@ local function on_player_joined_game(event)
 	Team_manager.draw_top_toggle_button(player)
 end
 
+local function on_gui_closed(event)
+	local player = game.players[event.player_index]
+	local trusted = Session.get_trusted_table()
+	if event.entity == nil or event.entity.last_user == nil or event.entity.get_recipe() == nil then return end
+	
+	if not trusted[player.name]
+	and (event.entity.get_recipe().name == "fast-transport-belt" or event.entity.get_recipe().name == "fast-underground-belt" or event.entity.get_recipe().name == "fast-splitter")
```
Comment (TheBigZet) [inline maps/biter_battles_v2/main.lua:40]: Push the event.entity.get_recipe() outside the if to call it just once
Or
local belts = {"fast-transport-belt", "fast-underground-belt", "fast-splitter"}
if belts[event.entity.get_recipe().name] then ...

**Review by TheBigZet — COMMENTED** [review]
Does not prevent copying recipes.
Clears the recipe on gui_close even if the player wasn't the one who set it

Also, what are all this old PRs? try updating your fork

File: maps/biter_battles_v2/main.lua (hunk 1 above)
Comment (Ragnarok77-factorio) [reply maps/biter_battles_v2/main.lua]: Okay, added the belts variable as you did, just added =true otherwise lua NOK.

**Comment by Ragnarok77-factorio** [comment]
> Does not prevent copying recipes. Clears the recipe on gui_close even if the player wasn't the one who set it
> 
> Also, what are all this old PRs? try updating your fork

For old PRs, it was just a fail, I cloned from master instead of dev branch at the beginning, giving all the weird commits before, should have started from dev branch instead.
==
For recipes, I added this case then, with the event "on_entity_settings_pasted". So this one should be OK.

For clearing the recipe, I don't get it, I tested with 2 players, if the untrusted opens a red belt asm gui, and close the gui, it doesn't clear the recipe (due to this check in code : entity.last_user.name == player.name)

I bet you had this issue because you trusted yourself, then you change recipe to red belt, then you untrusted yourself, then you opened gui of red belt asm and closed it, and it cleared it since last user is you (which is normal since you are untrusted, no red belt recipe for untrusted)
@TheBigZet

**Comment by Ragnarok77-factorio** [comment]
For trace, running away from asm with asm gui opened is not an exploit, as the event is called even in this case when the gui closes itself.

**Comment by developer-8** [comment]
Can confirm copying recipes with ctrl+c works

**Comment by Ragnarok77-factorio** [comment]
Yeah, still a workaround. I was doing it with shift (which doesn't work) but with ctrl c, indeed. And needs to think about bots too depending on how it's made.

But I'm not seeing event for that ctrl c when it's not a new building you are making (calling event built), but when you use it to change a recipe of existing asm.. Pain

**Comment by Ragnarok77-factorio** [comment]
I think I solved it now with cursor, however the solution is quite hacky but can't find better. @developer-8 @TheBigZet

**Comment by developer-8** [comment]
Alt+B take blueprint, then hold Shift while selecting assembler with recipe allow copy and paste it

**Comment by Ragnarok77-factorio** [comment]
> Alt+B take blueprint, then hold Shift while selecting assembler with recipe allow copy and paste it

Okay, you are right, alt b doesn't work and it's blocked, but alt b then hold shift works. Would never have found that lol. Let's see if if I can find something to handle this case too. Thanks !

**Comment by Ragnarok77-factorio** [comment]
Needs also to update it to remove hacky way and use get_blueprint_entities() with . recipe instead.

**Comment by Ragnarok77-factorio** [comment]
Waiting for clear cursor bugfix in 1.1.66 so it doesnt crash the game anymore and I can continue on this feature (very unlucky to get this crash factorio bug, glad they fixed it just 1 day after report, very good devs)

**Comment by Ragnarok77-factorio** [comment]
Okay ,server updated, so I could finish the PR, ready for review !

File: maps/biter_battles_v2/main.lua [hunk 2]
```diff
@@ -349,19 +368,56 @@ Event.add_event_filter(defines.events.on_post_entity_died, {
...
 	type = "unit",
 })
+
+local function on_entity_settings_pasted(event)
+	if event.destination == nil then return end
+	clear_recipe_asm_if_redbelt_and_untrusted(event.destination,event.player_index,true)
+end
+
+local function prevent_player_paste_prohibited_items_when_untrusted(playerIndex)
+	local player = game.players[playerIndex]
```
Comment (TheBigZet) [inline maps/biter_battles_v2/main.lua:378]:
```suggestion
	local player = game.get_player(playerIndex)
```

File: maps/biter_battles_v2/main.lua [hunk 3]
```diff
@@ -32,6 +33,24 @@ local function on_player_joined_game(event)
 	Team_manager.draw_top_toggle_button(player)
 end
 
+local function clear_recipe_asm_if_redbelt_and_untrusted(entity, playerIndex,cloneCase)
+	local player = game.players[playerIndex]
```
Comment (TheBigZet) [inline maps/biter_battles_v2/main.lua:37]:
```suggestion
	local player = game.get_player(playerIndex)
```

**Comment by developer-8** [comment]
The more you fix it, the more obscure the way around I will find.
1. Open assembler and chose red belt recipe
2. Kill assembler without closing it window. The assembler ghost will preserve the recipe
3. Place new assembler over ghost.

**Comment by Ragnarok77-factorio** [comment]
Resolving conflicts first due to branch master that evolved/new target branch.

**Comment by Ragnarok77-factorio** [comment]
> The more you fix it, the more obscure the way around I will find.
> 
>     1. Open assembler and chose red belt recipe
> 
>     2. Kill assembler without closing it window. The assembler ghost will preserve the recipe
> 
>     3. Place new assembler over ghost.

Fixed now I think ?

The more I add edge cases though, the more I'm scared of breaking something.

**Comment by developer-8** [comment]
1. Open assembler and chose red belt recipe
2. Kill assembler without closing it window. The assembler ghost will preserve the recipe
3. Bots place new assembler.

**Comment by developer-8** [comment]
1. Open assembler and chose red belt recipe
2. Kill yourself without closing it window. The assembler preserve the recipe

**Comment by Ragnarok77-factorio** [comment]
> 1. Open assembler and chose red belt recipe
> 
>     2. Kill assembler without closing it window. The assembler ghost will preserve the recipe
> 
>     3. Bots place new assembler.

This one, fixed I think.

For the one with suicide, didnt look yet but not even sure it's easily fixable (hope on death you get info on gui or something)

**Comment by TheBigZet** [comment]
Maybe it's time to admit it's impossible to implement

**Comment by developer-8** [comment]
1. Open assembler and chose red belt recipe
2. Make blueprint of assembler without closing it window. The blueprint has assembler with red belt recipe.
3. Put that blueprint in a book
4. Place the book as blueprint over assembler.

**Comment by Ragnarok77-factorio** [comment]
Ahah, yeah this PR is a pain (with trying not to impact perf, so no on tick function for example). But to be honest, if someone really knows this kind of workarounds, he's probably not a new player and knows how to use redbelt, or he's a griefer to know these exploits but then he will just get banned whenever he will grief main.

I feel like what 99,9% of players will try to do for red belts is now resolved in the PR (which was the goal of the poll : stop new players setting red belts asm). I will stop updating PR to fix these last exploits. 

Whatever you feel like for this PR, if you want 0 exploit, then you can close PR and we can just say in the poll => not implemented due to presence of some exploits by experimented players not trusted. 

Or we test it in live game then merge it the way it is now (and maybe one day we will fix the exploits) .

**Comment by Ragnarok77-factorio** [comment]
Closing it, if someone wanna fix conflicts and update it, feel free, I wont maintain it

**Pull request closed without merging**